### ISSUE TITLE
The pretrained weight of vgg

### ISSUE THREAD

**Issue by puhan123** [ZiweiWangTHU/BiDet#18]
When you pre-train the vgg on ImageNet, Do you add BN or shortcut(SC)?

If you don't add it, when you  load the pretrained model on BiDet,  Will the parameters not match?

**Comment by Wuziyi616** [comment]
Hi! Yes, I do add the BN and shortcut when pretraining VGG on the ImageNet. In other words, the VGG used in pretraining and detection task training have the **same** architecture. Also, you can refer to this issue #9 for more details about VGG pretraining.

**Comment by puhan123** [comment]
Thank you for your reply!

When you pre-train the vgg, did you also clip the relu function of some layers(like bidet)?

**Comment by Wuziyi616** [comment]
Yes, in the pretraining stage, we also train a **binary VGG16**. That is to say, the pretrained VGG16 is completely the same as the backbone VGG16 used in BiDet in terms of architecture, binarization, etc. To make it more clear, I adopt the same implementation as in BiReal-Net paper, which first trains a full-precision model, then use tanh activation to replace relu and finally use binarization activation to finetune. You can refer to their paper for more details.

**Comment by Wuziyi616** [comment]
Oh, I need to revise what I have said. The pretrained VGG and the backbone VGG is not completely the same. Since **pretraining is done in the classification task**, so I use a global average pooling after the conv layers to reduce feature dimension, followed by a 1000-way fully connected layer to output class probability. And when use the pretrained VGG as backbone, we simply drop the global average pooling layer and the fully connected layer. But anyway I think you can just use the pretrained VGG weight I provided :)

**Comment by puhan123** [comment]
Thank you very much for your timely reply! It helped me a lot!

**Comment by puhan123** [comment]
Excuse me, When you train the binary vgg_16 in the classification task, How many training epochs have you done? What is the final training accuracy? 

Because I train the vgg_16 using a different binary method now, and I find it very time consuming. I want to make sure if I need to train to the best accuracy?

Thank you very much.

**Comment by Wuziyi616** [comment]
@puhan123 Yes very good question. The training of VGG on ImageNet is indeed really really slow. I forgot about the exact epoch number for training. If I remember correctly, I used 4 GPUs, the batch size may be 4*64=256, and I trained for several days using Adam optimizer with initial lr=1e-3 and decayed twice to 1e-4 and 1e-5 respectively. The final Top-1 accuracy is around 65%.

Actually, I think it should be very difficult if you train the binary VGG from scratch, so I'd suggest you to adopt the training strategy as in Bi-Real-Net. Also, Adam optimizer is much better than SGD for binary neural networks. And a big batch size is useful to stabilize the training.
